File: code/backend/Odin/src/main/kotlin/pt/isel/odin/service/interfaces/ModuleService.kt
package pt.isel.odin.service.interfaces

import pt.isel.odin.model.Module

/**
 * Service for Modules.
 */
interface ModuleService {

    /**
     * Gets a Module by its id.
     *
     * @param id the Module id.
     */
    fun getById(id: Long): Module

    /**
     * Gets all Modules.
     */
    fun getAll(): List<Module>

    /**
     * Saves or Updates a Module.
     *
     * @param moduleRequest the Module to save.
     */
    fun save(moduleRequest: Module): Module

    /**
     * Deletes a Module by its id.
     *
     * @param id the Module id.
     */
    fun delete(id: Long)
}

File: code/frontend/Tyr/src/services/module/ModuleService.ts
import {apiRequest} from '../../axios/apiRequest';
import {CreationModuleResult, DeleteModuleResult, GetAllModulesResult, GetModuleResult} from "./ModuleResult";
import {ModuleInputModel} from "./models/ModuleInputModel";

/**
 * Service to manage modules
 */
export namespace ModuleService {

    const basePath = '/modules';

    /**
     * Get module by id
     *
     * @param id the id of the module
     *
     * @returns the module
     */
    export async function getById(id: number): Promise<GetModuleResult> {
        return apiRequest('get', `${basePath}/${id}`);
    }

    /**
     * Get all modules
     *
     * @returns all modules
     */
    export async function getAll(): Promise<GetAllModulesResult> {
        return apiRequest('get', basePath);
    }

    /**
     * Save a module
     *
     * @param moduleRequest the module to save
     *
     * @returns the result of the creation
     */
    export async function save(moduleRequest: ModuleInputModel): Promise<CreationModuleResult> {
        return apiRequest('post', `${basePath}/save`, moduleRequest);
    }

    /**
     * Update a module
     *
     * @param moduleRequest the module to update
     *
     * @returns the result of the update
     */
    export async function update(moduleRequest: ModuleInputModel): Promise<CreationModuleResult> {
        return apiRequest('put', `${basePath}/update`, moduleRequest);
    }

    /**
     * Delete a module by id
     *
     * @param id the id of the module
     *
     * @returns the result of the deletion
     */
    export async function deleteById(id: number): Promise<DeleteModuleResult> {
        return apiRequest('delete', `${basePath}/${id}`);
    }
}
